Issue 1795 - RFE - Enable logging for libldap and libber in error log

droideck opened the pull request:
Description: Libraries like libldap, libber do error and debug
logging, but it is not available in the DS logs.

Provide a way to enable the third party logging in DS.
Add nsslapd-external-libs-debug-enabled attribute to 'cn=config'
which will enable all of the levels abailable in libldap and libber.
The setting should be used only for debugging purposes as
it prints all of the operataions with a great verbosity.

The code for log_external_libs_debug_print() and
log_external_libs_debug_set_log_fn() functions are provided
by a former Red Hat employee - Ludwig Krispenz.

Fixes: #1795

Reviewed by: ?

Firstyear commented on ldap/servers/slapd/log.c:
```diff
@@ -441,6 +441,23 @@ log_set_logging(const char *attrname, char *value, int logtype, char *errorbuf,
+void
+log_external_libs_debug_print(char *buffer)
+{
+    slapi_log_error(SLAPI_LOG_FATAL, "OpenLDAP Client", "%s", buffer);
```
Shouldn't this be "external lib -> " rather than "openldap" since we could have many types of external lib?

Firstyear replied:
Log level should be TRACE or DEBUG maybe too?

droideck replied:
> Shouldn't this be "external lib -> " rather than "openldap" since we could have many types of external lib?

If we won't mention the library name then log messages like this one won't make sense:

```
[07/Dec/2020:10:43:58.042403434 -0500] - ERR - OpenLDAP Client - attempting to connect:
```

But I think you have a good point and I should give this function a more accurate name - `log_external_libs_debug_print` - probably something like `log_external_libs_debug_openldap_print`?

> Log level should be TRACE or DEBUG maybe too?

My idea of the usage is like this:
1. You have an issue;
2. You set the debug level you need (maybe only replication?);
3. You think that the problem happens somewhere in the middle of replica auth;
4. You enable nsslapd-external-libs-debug-enabled attribute;
5. You see only OpenLDAP and Replication level output (you don't need any other noise from TRACE or DEBUG)

The same way you can 'inject' the info in any other debug level.

So I think it makes sense to have it distinctive. So it's the reason I put it to the minimal level (FATAL).
As an option, we can put it to the separate loglevel, maybe?
What do you think?

Firstyear replied:
well I think fatal is just "too intense" - I think our default levels always display warn and error, so it could be better to set this to warn instead. 

But yes, now having read the other comments I agreed that the name should be "libldap/libber" instead of openldapclient in this, and that the function should be renamed to suit :)

Firstyear commented on ldap/servers/slapd/log.c:
```diff
@@ -441,6 +441,23 @@ log_set_logging(const char *attrname, char *value, int logtype, char *errorbuf,
+int
+log_external_libs_debug_set_log_fn(void)
+{
+    int rc = ber_set_option(NULL, LBER_OPT_LOG_PRINT_FN, log_external_libs_debug_print);
```
How do you s-et the print fn for ldap? or does it just use stderr?

droideck replied:
It uses the LBER_OPT_LOG_PRINT_FN so it is set only one time (if I understand it correctly)
https://git.openldap.org/openldap/openldap/-/blob/master/libraries/libldap/print.c#L60

Firstyear commented on ldap/servers/slapd/daemon.c:
```diff
@@ -493,6 +499,13 @@ disk_monitoring_thread(void *nothing __attribute__((unused)))
                         config_set_auditfaillog_enabled(LOGGING_ON);
                     }
+                    if (using_external_libs_debug) {
+                        if (config_set_external_libs_debug_enabled(CONFIG_EXTERNAL_LIBS_DEBUG_ENABLED,
```
What's the intent for these calls to set the external lib debug enabled? Is it so you don't have to enable it via dse.ldif?

droideck replied:
It is for disk_monitoring daemon only.
It disables the logging setting if the disk threshold was crossed (and it will set it back to the previous state after the space is available once again).

Firstyear replied:
OHHHHHh I see. That's okay then :)

Firstyear commented on dirsrvtests/tests/suites/disk_monitoring/disk_monitoring_test.py:
```diff
@@ -731,6 +731,38 @@ def test_valid_operations_are_permitted(topo, setup, reset_logs):
     assert topo.standalone.config.set('nsslapd-disk-monitoring-logging-critical', 'on')
 
 
+def test_enable_external_libs_debug_log(topo, setup, reset_logs):
```
IIRC this test should have a marker or warning of some kind about how it will use disk space. Also some filesystems will see a dev/zero dd and just make a sparse file and can over-allocate ... there are also risks with corruption on some fs's (btrfs famous enospc issues).

droideck replied:
Good catch! Fixed:)

tbordaz commented on ldap/servers/slapd/log.c:
```diff
@@ -441,6 +441,23 @@ log_set_logging(const char *attrname, char *value, int logtype, char *errorbuf,
     return LDAP_SUCCESS;
 }
 
+void
```
I think it could be defined as static

tbordaz reviewed (approved):
Except this minor comment, the patch looks good to me. ACK

droideck committed:
Issue 1795 - RFE - Enable logging for libldap and libber in error log

Description: Libraries like libldap, libber do error and debug
logging, but it is not available in the DS logs.

Provide a way to enable the third party logging in DS.
Add nsslapd-external-libs-debug-enabled attribute to 'cn=config'
which will enable all of the levels abailable in libldap and libber.
The setting should be used only for debugging purposes as
it prints all of the operataions with a great verbosity.

The code for log_external_libs_debug_print() and
log_external_libs_debug_set_log_fn() functions are provided
by a former Red Hat employee - Ludwig Krispenz.

Fixes: #1795

Reviewed by: @Firstyear and @tbordaz (Thanks!)
